File: src/exercises/chapter1/lesson7.rb
# frozen_string_literal: true

# S(n) = ½ + 2/3 + ¾ + …. + n / n + 1
# @param {Integer} n - The upper limit of the range (positive integer)
# @return {Float} - The sum of the series up to n / (n + 1)

def solution(n)
  result = 0.0
  (1..n).each do |i|
    result += i.to_f / (i + 1) # Calculate each element of the array and add it to the result
  end
  result
end

File: src/exercises/chapter1/Lesson7.ts
/**
 * Lesson 7: Calculate S(n) = 1/2 + 2/3 + 3/4 + … + n/(n + 1)
 */
export default class Lesson7 {

  /**
   * Calculates the sum S(n) = 1/2 + 2/3 + 3/4 + … + n/(n + 1).
   * @param {number} n - The upper limit number to calculate the sum for.
   * @return {number} - The sum of the series for the given n.
   */
  solution(n: number): number {
    // Initialize result to store the sum
    let result = 0

    // Loop through all numbers from 1 to n and accumulate the sum of i / (i + 1)
    for (let i = 1; i <= n; i++) {
      // Add the fraction i / (i + 1) to the result
      result += i / (i + 1)
    }

    // Return the calculated sum
    return result
  }
}
